### PR TITLE
#9336: Refactoring moreh layernorm

### PR DESCRIPTION
**Pull request by hschoi4448**
### Ticket
- https://github.com/tenstorrent/tt-metal/issues/9336

### Problem description
- The current `moreh_layernorm` only accepts 4D tensors as input. To meet this restriction, the `mean` and `rstd` tensors are stored inefficiently, with only one valid data point per tile.
- the current implementation does not support `optional outputs` and the `fp32_dest_acc_en` feature.

### What's changed
- refactoring moreh_layernorm
  - add fp32_dest_acc_en support
  - support non 4d tensor
  - support optional output tensor

### Checklist
- [x] Post commit CI passes
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by sangwon-chae** [comment]
@TT-BrianLiu @tt-aho @mywoodstock Would you please review this PR? Code owners review is required. Thanks.

File: tt_eager/tt_lib/csrc/operations/primary/module.hpp
Comment (ayerofieiev-tt) [inline tt_eager/tt_lib/csrc/operations/primary/module.hpp:760]: same as here

https://github.com/tenstorrent/tt-metal/pull/9804#discussion_r1661859299

File: tt_eager/tt_dnn/op_library/moreh_layernorm_backward/moreh_layernorm_backward_op.cpp
Comment (ayerofieiev-tt) [inline tt_eager/tt_dnn/op_library/moreh_layernorm_backward/moreh_layernorm_backward_op.cpp:147]: consider
```suggestion
    TT_FATAL(false, "Create output tensor is not supported yet. Fix this after the #9552 issue is addressed.");
```

File: tt_eager/tt_dnn/op_library/moreh_layernorm_backward/moreh_layernorm_backward_op.cpp
Comment (ayerofieiev-tt) [inline tt_eager/tt_dnn/op_library/moreh_layernorm_backward/moreh_layernorm_backward_op.cpp:136]: use tt_fatal here too

**Review by ayerofieiev-tt — APPROVED** [review]
lgtm. 
would be great to get another review from someone more knowledgable with kernels

File: tt_eager/tt_dnn/op_library/moreh_layernorm_backward/moreh_layernorm_backward_op.cpp
Comment (hschoi4448) [reply tt_eager/tt_dnn/op_library/moreh_layernorm_backward/moreh_layernorm_backward_op.cpp:147]:
```C++
TT_FATAL(false, "Create output is not supported yet. Fix this after the #9552 issue is addressed.");
```
I have made the changes as you suggested.

File: tt_eager/tt_dnn/op_library/moreh_layernorm_backward/moreh_layernorm_backward_op.cpp
Comment (hschoi4448) [reply tt_eager/tt_dnn/op_library/moreh_layernorm_backward/moreh_layernorm_backward_op.cpp:136]:
```C++
T_FATAL(false, "The compute_output_shapes function in MorehLayerNormBackwardGammaBetaGrad is not implemented.");
```
I have made the changes as shown in the code above.

File: tt_eager/tt_lib/csrc/operations/primary/module.hpp
Comment (hschoi4448) [reply tt_eager/tt_lib/csrc/operations/primary/module.hpp]:
```C++
tt_eager/tt_lib/csrc/operations/primary/module.hpp

py::arg("output_mem_config").noconvert() = operation::DEFAULT_OUTPUT_MEMORY_CONFIG,
->
py::arg("memory_config").noconvert() = std::nullopt,

```
https://github.com/tenstorrent/tt-metal/pull/9636/commits/b2168abdbefbb42af71a7c44cb5cbae4310e2bd3#diff-635a2960d38f45b47d93f3be3e4cc45b69c32e0e904f69d20486a47ee7803b79

I have modified it to receive `memory_config` like other implementations: `const std::optional<MemoryConfig>& memory_config = std::nullopt`.

**Review by TT-BrianLiu — APPROVED** [review]

**Pull request merged**
